include/lug/System/Memory.hpp

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include <cstddef>
+#include <type_traits>
+#include <new>
+#include <memory>
+
+#include <lug/System/Memory/Arena.hpp>
+#include <lug/System/Memory/Policies/Thread.hpp>
+#include <lug/System/Memory/Policies/BoundsChecker.hpp>
+#include <lug/System/Memory/Policies/MemoryMarker.hpp>
+
+
+// The variadic arguments are always "arena, args..." (args only work for non array allocation and array of non pod)
+#define LUG_NEW_ALIGN(T, alignment, ...) ::lug::System::Memory::new_one<T>(alignment, __FILE__, __LINE__, __VA_ARGS__)
+#define LUG_NEW_ARRAY_ALIGN(T, alignment, ...) LUG_NEW_ARRAY_ALIGN_SIZE(typename std::remove_all_extents<T>::type, alignment, std::extent<T>::value, __VA_ARGS__)
+#define LUG_NEW_ARRAY_ALIGN_SIZE(T, alignment, size, ...) ::lug::System::Memory::new_array<typename std::remove_all_extents<T>::type>(alignment, size, __FILE__, __LINE__, __VA_ARGS__)
+
+#define LUG_NEW(T, ...) LUG_NEW_ALIGN(T, alignof(T), __VA_ARGS__)
+#define LUG_DELETE(object, arena) ::lug::System::Memory::delete_one(object, arena)
+
+#define LUG_NEW_ARRAY(T, ...) LUG_NEW_ARRAY_ALIGN(T, alignof(typename std::remove_all_extents<T>::type), __VA_ARGS__)
+#define LUG_NEW_ARRAY_SIZE(T, size, ...) LUG_NEW_ARRAY_ALIGN_SIZE(T, alignof(typename std::remove_all_extents<T>::type), size, __VA_ARGS__)
+#define LUG_DELETE_ARRAY(object, arena) ::lug::System::Memory::delete_array(object, arena)
+
+
+namespace lug {
+namespace System {
+namespace Memory {
+
+template <typename T, class Arena, class ...Args>
+T* new_one(size_t alignment, const char* file, size_t line, Arena& arena, Args&&... args);
+
+template <typename T, class Arena>
+void delete_one(T* object, Arena& arena);
+
+template <typename T, class Arena, class ...Args, typename std::enable_if<!std::is_pod<T>::value, int>::type = 0>
+T* new_array(size_t alignment, size_t nb, const char* file, size_t line, Arena& arena, Args&&... args);
+
+template <typename T, class Arena, typename std::enable_if<!std::is_pod<T>::value, int>::type = 0>
+void delete_array(T* ptr, Arena& arena);
+
+template <typename T, class Arena, typename std::enable_if<std::is_pod<T>::value, int>::type = 0>
+T* new_array(size_t alignment, size_t nb, const char* file, size_t line, Arena& arena);
+
+template <typename T, class Arena, class ...Args, typename std::enable_if<std::is_pod<T>::value, int>::type = 0>
+T* new_array(size_t alignment, size_t nb, const char* file, size_t line, Arena& arena, Args&&... args) = delete;
+
+template <typename T, class Arena, typename std::enable_if<std::is_pod<T>::value, int>::type = 0>
+void delete_array(T* ptr, Arena& arena);
+
+// unique_ptr
+namespace priv {
+
+template <typename T>
+struct make_unique_deleter {
+    using Deleter = std::function<void (T*)>;
+};
+
+template <typename T>
+struct make_unique_deleter<T[]> {
+    using Deleter = std::function<void (T*)>;
+};
+
+template <typename T, size_t Count>
+struct make_unique_deleter<T[Count]> {
+    using Deleter = std::function<void (T*)>;
+};
+
+}
```

Comment:
Add `// priv` to indicate that you are closing the namespace. Apply for all files in the review. It's difficult without indentation for a reader to understand quickly what are these brackets for.